lib/charms/openstack_libs/v0/rabbitmq_requires.py

```diff
@@ -0,0 +1,206 @@
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class RabbitMQConnectedEvent(EventBase):
+    """RabbitMQ connected Event."""
+
+    pass
+
+
+class RabbitMQReadyEvent(EventBase):
+    """RabbitMQ ready for use Event."""
+
+    pass
+
+
+class RabbitMQGoneAwayEvent(EventBase):
+    """RabbitMQ relation has gone-away Event."""
+
+    pass
+
+
+class RabbitMQServerEvents(ObjectEvents):
+    """Events class for `on`."""
+
+    connected = EventSource(RabbitMQConnectedEvent)
+    ready = EventSource(RabbitMQReadyEvent)
+    goneaway = EventSource(RabbitMQGoneAwayEvent)
+
+
+class RabbitMQRequires(Object):
+    """Requires side interface for rabbitmq interface type."""
+
+    on = RabbitMQServerEvents()
+
+    def __init__(self, charm, relation_name: str, username: str, vhost: str):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.username = username
+        self.vhost = vhost
+
+        self.framework.observe(
+            self.charm.on[relation_name].relation_joined,
+            self._on_rabbitmq_relation_joined,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed,
+            self._on_rabbitmq_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_departed,
+            self._on_rabbitmq_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken,
+            self._on_rabbitmq_relation_broken,
+        )
+
+    def _on_rabbitmq_relation_joined(self, event):
+        """Rabbitmq relation joined."""
+        logging.debug("RabbitMQ on_joined")
+        self.on.connected.emit()
+        self.register()
+
+    def _on_rabbitmq_relation_changed(self, event):
+        """Rabbitmq relation changed."""
+        logging.debug("RabbitMQ on_changed")
+        try:
+            self.password
+            self.on.ready.emit()
+        except KeyError:
+            pass
+
+    def _on_rabbitmq_relation_broken(self, event):
+        """Rabbitmq relation broken."""
+        logging.debug("RabbitMQ on_broken")
+        self.on.goneaway.emit()
+
+    @property
+    def _rabbitmq_rel(self) -> Relation:
+        """The RabbitMQ relation."""
+        return self.framework.model.get_relation(self.relation_name)
+
+    @property
+    def hostname(self) -> str:
+        """Return the hostname."""
+        return self.get_data("hostname")
+
+    @property
+    def password(self) -> str:
+        """Return the password."""
+        return self.get_data("password")
+
+    def get_data(self, key: str) -> str:
+        """Return the value for the given key."""
+        for unit in self._rabbitmq_rel.units:
+            data = self._rabbitmq_rel.data[unit]
+            try:
+                return data[key]
+            except KeyError:
+                return None
```

Comment:
This pattern has the effect of only checking the first unit in the relation; there is a similar issue in the gnocchi relation.

It will work but only when the first unit in the list provides the value.

If you just pass the exception handler then None will get returned by default if the value is not actually found anywhere on any unit on the relation.